src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+    source_root_config: SourceRootConfig
+) -> AncestorInitPyFiles:
+  """Find any ancestor __init__.py files for the given targets.
+
+  Includes sibling __init__.py files. Returns the files stripped of their source roots.
+  """
+  source_roots = source_root_config.get_source_roots()
+  # Find the ancestors of all dirs containing .py files, including those dirs themselves.
+  source_dir_ancestors: Set[Tuple[str, str]] = set()  # Items are (src_root, path incl. src_root).
+  sources_snapshots = await MultiGet(
+    Get[Snapshot](Digest, target.adaptor.sources.snapshot.directory_digest)
+    for target in targets if isinstance(target.adaptor, PythonTargetAdaptor)
+  )
+  for sources_snapshot in sources_snapshots:
+    for file in sources_snapshot.files:
+      source_dir_ancestor = os.path.dirname(file)
+      source_root = source_root_or_raise(source_roots, file)
+      # Do not allow the repository root to leak (i.e., '.' should not be a package in setup.py).
+      while source_dir_ancestor != source_root:
+        source_dir_ancestors.add((source_root, source_dir_ancestor))
+        source_dir_ancestor = os.path.dirname(source_dir_ancestor)
+
+  source_dir_ancestors_list = list(source_dir_ancestors)  # To force a consistent order.
+
+  # Note that we must MultiGet single globs instead of a a single Get for all the globs, because
+  # we match each result to its originating glob (see use of zip below).
+  ancestor_init_py_snapshots = await MultiGet[Snapshot](
+    Get[Snapshot](PathGlobs,
+                  PathGlobs(include=(os.path.join(source_dir_ancestor[1], '__init__.py'),)))
+    for source_dir_ancestor in source_dir_ancestors_list
+  )
+
+  source_root_stripped_ancestor_init_pys = await MultiGet[Digest](
+    Get[Digest](DirectoryWithPrefixToStrip(
+      directory_digest=snapshot.directory_digest, prefix=source_dir_ancestor[0])
+  ) for snapshot, source_dir_ancestor in zip(ancestor_init_py_snapshots, source_dir_ancestors_list))
+
+  return AncestorInitPyFiles(source_root_stripped_ancestor_init_pys)
+
+
+def _is_exported(target: HydratedTarget) -> bool:
+  return getattr(target.adaptor, 'provides', None) is not None
+
+
+@rule(name="Get requirements")
+async def get_requirements(dep_owner: DependencyOwner) -> ExportedTargetRequirements:
+  tht = await Get[TransitiveHydratedTargets](
+    BuildFileAddresses([dep_owner.exported_target.hydrated_target.address]))
+
+  ownable_tgts = [tgt for tgt in tht.closure
+                  if isinstance(tgt.adaptor, (PythonTargetAdaptor, ResourcesAdaptor))]
+  owners = await MultiGet(Get[ExportedTarget](OwnedDependency(ht)) for ht in ownable_tgts)
+  owned_by_us = set()
+  owned_by_others = set()
+  for tgt, owner in zip(ownable_tgts, owners):
+    (owned_by_us if owner == dep_owner.exported_target else owned_by_others).add(tgt)
+
+  # Get all 3rdparty deps of our owned deps.
+  #
+  # Note that we need only consider requirements that are direct dependencies of our owned deps:
```

Comment:
Hm. "strict deps" are not generally a requirement within Pants python codebases... is this a change in behaviour, or is this how v1 worked in this case?

Reply:
This doesn't actually require strict deps, and is how v1 worked:  

Say exported target C depends on exported target A and regular target B.  In other words, C's dist publishes the code from C and B, and A's dist publishes the code from A. 

So we do want C and B's 3rdparty requirements to become the requirements of C's dist, but we do *not* want A's requirements to (which is why we don't just take all transitive deps). So if C references a 3rdparty symbol that it depends on only via A, distC won't list that requirement (but distA will, and distC will list distA, so the non-strict effective dep will be preserved). If you want your dists to have strict deps, you need your targets to as well, at least across exported target boundaries.